pkg/controller.v1/paddlepaddle/envvar.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2021 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+package paddle
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	commonv1 "github.com/kubeflow/common/pkg/apis/common/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	kubeflowv1 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v1"
+)
+
+const (
+	EnvMasterEndpoint = "PADDLE_MASTER"
+	EnvNumNodes       = "PADDLE_NNODES"
+	EnvJobID          = "PADDLE_JOB_ID"
+	EnvServerNum      = "PADDLE_SERVER_NUM"
+	EnvTrainerNum     = "PADDLE_TRAINER_NUM"
+)
+
+// EnvVarGenerator is the environment variable generator interface.
+type EnvVarGenerator interface {
+	Generate(job *kubeflowv1.PaddleJob) ([]corev1.EnvVar, error)
+}
+
+func setPodEnv(obj interface{}, podTemplateSpec *corev1.PodTemplateSpec, rtype, index string) error {
+	paddlejob, ok := obj.(*kubeflowv1.PaddleJob)
+	if !ok {
+		return fmt.Errorf("%+v is not a type of PaddleJob", obj)
+	}
+
+	rank, err := strconv.Atoi(index)
+	if err != nil {
+		return err
+	}
+
+	totalReplicas := getTotalReplicas(paddlejob)
+
+	for i := range podTemplateSpec.Spec.Containers {
+		// Initialize the environment variables.
+		if len(podTemplateSpec.Spec.Containers[i].Env) == 0 {
+			podTemplateSpec.Spec.Containers[i].Env = make([]corev1.EnvVar, 0)
+		}
+		// Set PYTHONUNBUFFERED to true, to disable output buffering.
+		// Ref https://stackoverflow.com/questions/59812009/what-is-the-use-of-pythonunbuffered-in-docker-file.
+		podTemplateSpec.Spec.Containers[i].Env = append(podTemplateSpec.Spec.Containers[i].Env, corev1.EnvVar{
+			Name:  "PYTHONUNBUFFERED",
+			Value: "0",
+		})
```

Comment:
Is this a specific requirement for paddle pods?

Reply:
Yes, it almost necessary for paddle training in cloud environment, otherwise, the log is buffered which is confused for user check the training process.